### PR TITLE
Csm Inati 2D and 3D

### LINKED ISSUE #64: Csm inati method
could be implemented via a wrapper for https://github.com/ismrmrd/ismrmrd-python-tools/blob/17fe6fbf1bb645112e2ad0023eb4f42afb36421e/ismrmrdtools/coils.py#L92

### PR DESCRIPTION
**Pull request by ckolbPTB**
supersedes #173 
closes #64 

Too many things changes since we last worked on this so I started a fresh PR.

### CODE REVIEW

**Comment by ckolbPTB** [comment]
I verified the result with the output of the old PtbPyRecon inati code for several examples.

**Comment by ckolbPTB** [comment]
One last problem is using `torch.is_close`:
UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::isclose. 

Maybe somebody with more pytorch-knowledge could comment if this is a problem and we should change this or accept the warning in the tests.

File: src/mrpro/data/CsmData.py [hunk 1]
```diff
@@ -53,6 +54,31 @@ def from_idata_walsh(
...
+        ----------
+        idata
+            IData object containing the images for each coil element.
+        smoothing_width
+            Size of the smoothing kernel.
+        """
+        from mrpro.algorithms.csm.inati import inati
+
+        # convert smoothing_width to SpatialDimension if int
+        if isinstance(smoothing_width, int):
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]: why?

inati also accepts int.

**Review by fzimmermann89 — APPROVED** [review]
I am wondering if in the end it might be smarter to just set the number of iterations to a fixed value of say 10 and use vmap? 

Would this be faster if run for say a 32 image batch on GPU?
In case we ever want to use this during a NN training..

Could you test the performance difference @ckolbPTB or should I do it?
It might be either not worth it are a large difference, I really do not know.

Why is the relative change currently set to 0.1 (this seems high?) ? how many iterations would this corrpsopond to on some real data example?

File: src/mrpro/algorithms/csm/inati.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    if any(ks % 2 != 1 for ks in [smoothing_width.z, smoothing_width.y, smoothing_width.x]):
+        raise ValueError('kernel_size must be odd')
+
+    ks_halved = [ks // 2 for ks in smoothing_width.zyx]
+    padded_coil_images = torch.nn.functional.pad(
+        coil_images,
+        (ks_halved[-1], ks_halved[-1], ks_halved[-2], ks_halved[-2], ks_halved[-3], ks_halved[-3]),
+        mode=padding_mode,
+    )
+    # Get the voxels in an ROI defined by the smoothing_width around each voxel leading to shape (coils z y x ks*ks)
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/csm/inati.py]: this comment might be wrong? is it ks*ks`?  (or only in 2d)?

**Comment by fzimmermann89** [comment]
2d images are now also replicate padded. this does not change the result, only makes it a bit less efficient? Or does it change the result?

**Comment by ckolbPTB** [comment]
> Why is the relative change currently set to 0.1 (this seems high?) ? how many iterations would this corrpsopond to on some real data example?

For my test data (2 phantoms + 1 in-vivo) this leads to around 30 iterations.

> I am wondering if in the end it might be smarter to just set the number of iterations to a fixed value of say 10 and use vmap?

I will check that

> Would this be faster if run for say a 32 image batch on GPU? In case we ever want to use this during a NN training..

For 2D cases, vmap leads to a significant speedup compared to smap (2-6 for n_batch=10). For 3D cases (even if it is just a few slices) there is no difference anymore between smap and vmap.

**Comment by ckolbPTB** [comment]
> I am wondering if in the end it might be smarter to just set the number of iterations to a fixed value of say 10 and use vmap?

10 seems to be a good value. For my test data there are only minimal changes in the csm in the background between 10 and 20 iterations.

File: src/mrpro/algorithms/csm/inati.py (hunk 2 above)
Comment (ckolbPTB) [reply src/mrpro/algorithms/csm/inati.py]: It is prod(smoothing_width)

**Comment by ckolbPTB** [comment]
Currently if `smoothing_width` is `int` then we pad + smooth also along the z direction even if the data is 2D. I guess we do not want that...

**Pull request merged**